### ISSUE TITLE
dash does not refresh its hash table when PATH is re-declared

### ISSUE THREAD

**Issue by agarzola** [tj/n#706]
The following line recommends re-declaring `PATH` to ensure the new location of Node.js is loaded in the current session.

https://github.com/tj/n/blob/d569e331820399ae5224b640f1d556c64a87729d/bin/n#L744

However, in some environments that might not suffice. I ran into this issue on Platform.sh, which uses [`dash`](https://man7.org/linux/man-pages/man1/dash.1.html). Interestingly enough, after installing Node with `n auto` (our project has an `.nvmrc` file specifying v14), running `which node` points to the location where we told `n` to install Node: `/app/.global/bin/node`. However, running `node --version` yielded v12. Simply invoking `export PATH="$PATH"` did not solve this issue.

As it turns out, dash’s implementation of the `hash` built-in command does not refresh the hash table where executables are stored for easy access. The fix suggested to us by Platform.sh support was to invoke `hash -r` instead. This worked: after invoking that, `node --version` yielded the version number we expect (the one installed via `n`).

I will submit a PR shortly that updates the line linked above to include a note about how to solve this issue in dash.

**Comment by shadowspawn** [comment]
I remember considering `hash -r` vs `PATH=$PATH` but can't find any notes I made at the time.

Related issues mentioned in the CHANGELOG: advice to reset command hash when node location changes (#170 #381 #451 #588)

**Comment by shadowspawn** [comment]
More documentation:
- https://hyperpolyglot.org/unix-shells#cmd-resolution
- https://man7.org/linux/man-pages/man1/dash.1.html
- https://pubs.opengroup.org/onlinepubs/9699919799/utilities/hash.html

The Open Group docs for hash say this, which is probably why I went for `PATH` over `hash` originally:

> The effects of hash -r can also be achieved portably by resetting the value of PATH ; in the simplest form, this can be:
>
> PATH="$PATH"

**Comment by agarzola** [comment]
That is true of Bash’s implementation of `hash`. Alas, that appears to not be the case in Dash’s implementation.

**Comment by shadowspawn** [comment]
Yes "portably" is not universally. 🙊

**Comment by shadowspawn** [comment]
After some experimentation...

```
bash: PATH=$PATH or hash -r
fish: no action required
ksh: hash -r
tcsh: rehash
zsh: PATH=$PATH or hash -r
dash: hash -r
csh: rehash
ash: hash -r
```

**Comment by agarzola** [comment]
Oh, wow. So how about I modify my PR thusly:

>To reset the command location hash, start a new shell. If starting a new shell is not an option, here's what you can do:
>- For bash and zsh, execute `PATH="$PATH"` or `hash -r`.
>- For csh and tcsh, execute `rehash`.
>- For ash, dash, and ksh, execute `hash -r`.

**Comment by shadowspawn** [comment]
I am disappointed with `PATH="$PATH"` since it only works in some shells and they both support `hash -r` anyway!

Moving the command to the front to make it easier to spot, how about:

```
Note: the node command changed location and the old location may be remembered in your current shell.
         old : /usr/local/bin/node
         new : /root/.n/bin/node
If "node --version" shows the old version then either start a new shell, or reset the location hash with:
hash -r  (for bash, zsh, ash, dash, and ksh)
rehash   (for csh and tcsh)
```

**Comment by agarzola** [comment]
Agreed, that’s much better. I’ll update my PR with that here shortly!

**Comment by agarzola** [comment]
@shadowspawn All done! PR is ready for review at your leisure.

**Comment by shadowspawn** [comment]
Improvements included in `n` v8.0.2
